### PR TITLE
Fix building of gtest with -fvisibility=hidden [#451]

### PR DESCRIPTION
**Pull request by d-meiser**
This fixes [#451]. I've signed the CLA.

### CODE REVIEW

**Comment by d-meiser** [comment]
Is there any interest in these changes? Do I need to clean something up?

**Comment by d-meiser** [comment]
Is this the right place to submit pull requests? readme suggests it is but perhaps that's out of date?

**Comment by BillyDonahue** [comment]
Yes this is the correct place. The queue is a little neglected right now.
Sorry.

**Comment by d-meiser** [comment]
No problem. Let me know if I can do anything to make it easier to review or merge this PR.

File: googletest/include/gtest/internal/gtest-port.h [hunk 1]
```diff
@@ -920,18 +920,22 @@ using ::std::tuple_size;
...
-# define GTEST_API_
-#endif
+#elif __GNUC__ >= 4 || defined(__clang__)
+# if GTEST_CREATE_SHARED_LIBRARY
+#  define GTEST_API_ __attribute__((visibility ("default")))
+# else
+#  define GTEST_API_
+# endif
+#else
+# ifndef GTEST_API_
```
Comment (BillyDonahue) [inline googletest/include/gtest/internal/gtest-port.h]: The ifelse chain is hard to follow. How about moving this fallback out of the entire staircase, leaving it as an unconditional cleanup after the last #endif ?

File: googletest/include/gtest/internal/gtest-port.h [hunk 2]
```diff
@@ -920,18 +920,22 @@ using ::std::tuple_size;
...
 #  define GTEST_API_ __declspec(dllexport)
 # endif
-
-#endif  // _MSC_VER
-
-#ifndef GTEST_API_
-# define GTEST_API_
-#endif
+#elif __GNUC__ >= 4 || defined(__clang__)
+# if GTEST_CREATE_SHARED_LIBRARY
```
Comment (BillyDonahue) [inline googletest/include/gtest/internal/gtest-port.h]: Maybe this __attribute would be useless if the visibility preset wasn't going to be "hidden".

File: googletest/include/gtest/internal/gtest-port.h (hunk 1 above)
Comment (d-meiser) [inline googletest/include/gtest/internal/gtest-port.h]: Great suggestion.

File: googletest/include/gtest/internal/gtest-port.h (hunk 2 above)
Comment (d-meiser) [inline googletest/include/gtest/internal/gtest-port.h]: I'm not sure I fully understand. Are you proposing to get rid of the `#if GTEST_CREATE_SHARED_LIBRARY` because the `__attribute__((visibility ("default")))` doesn't have an effect if we're building a static library? I think that would be a valid simplification. I'll try that.

**Comment by d-meiser** [comment]
Thanks for the feedback. I'll update the pull request, hopefully by tonight.

**Comment by d-meiser** [comment]
@BillyDonahue I've cleaned up the definition of GTEST_API_. Let me know if this is better. Thanks.
